packages/server/src/fhir/sql.ts

```diff
@@ -613,6 +621,116 @@ export class SelectQuery extends BaseQuery implements Expression {
   }
 }
 
+export class SelectPerReferenceQuery extends BaseQuery implements Expression {
+  readonly joins: Join[];
+  readonly orderBys: OrderBy[];
+  limit_: number;
+
+  readonly referenceColumn: Column;
+  references: string[];
+  readonly outerQuery: SelectQuery;
+  readonly referenceQuery: SelectQuery;
+  readonly resultsQuery: SelectQuery;
+  readonly resultsInnerQuery: SelectQuery;
+
+  constructor(tableName: string, referenceColumn: Column, references: string[]) {
+    super(tableName);
+    this.joins = [];
+    this.orderBys = [];
+    this.limit_ = 0;
+
+    this.referenceColumn = referenceColumn;
+    this.references = [...references];
+    this.referenceQuery = new SelectQuery(this.tableName);
+    this.resultsInnerQuery = new SelectQuery(this.tableName);
+    this.resultsQuery = new SelectQuery('ranked', this.resultsInnerQuery);
+    this.outerQuery = new SelectQuery('references', this.referenceQuery);
+
+    this.referenceQuery.column(
+      new Column(undefined, `DISTINCT "${this.referenceColumn.columnName}" AS "reference"`, true)
+    );
+
+    this.referenceQuery.where(this.referenceColumn, 'IN', references, 'string'); // 'string' doesn't matter here?
+
+    this.outerQuery.joins.push(new Join('JOIN LATERAL', this.resultsQuery, 'results', new Literal('true')));
+
+    this.resultsInnerQuery.where(referenceColumn, '=', new Column('references', 'reference'));
+  }
+
+  column(column: Column | string): this {
+    this.outerQuery.column(new Column('results', getColumnName(column)));
+    this.resultsQuery.column(getColumnName(column));
+    this.resultsInnerQuery.column(column);
+    return this;
+  }
+
+  where(column: Column | string, operator?: keyof typeof Operator, value?: any, type?: string): this {
+    this.referenceQuery.where(column, operator, value, type);
+    this.resultsInnerQuery.where(column, operator, value, type);
+    return this;
```

Comment:
@codyebberson Per your comments on security filters, they are propagated into the inner queries here.